### PR TITLE
fix: Fixed AI chat function

### PR DESCRIPTION
**Pull request by Fleurxxx**
English | [简体中文](https://github.com/opentiny/tiny-engine/blob/develop/.github/PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.zh-CN.md)

# PR

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our [Commit Message Guidelines](https://github.com/opentiny/tiny-engine/blob/develop/CONTRIBUTING.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [x] Built its own designer, fully self-validated

## PR Type

What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Background and solution

1. Issues

- The conversation cannot proceed normally, and the questions posed do not display responses.
- The response content has not been converted to markdown format.
- There is a lack of voice input functionality.

2. New Features, which require a feature description along with effect images.

This PR fixes the AI conversation functionality.

- A new popup for filling in the access token has been added; if the access token expires, users will need to re-enter it to resume normal conversation.
- Markdown conversion has been added to ensure that the conversation content returned is in markdown format.
- A new voice input feature has been added, which allows for communication through voice-to-text conversion.

### What is the current behavior?

Issue Number: N/A

### What is the new behavior?

1. New dialog

image

image

2. Repair session

image

3. Speech to text

image

4. prompt modification

录屏2024-07-27-20 16 56

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Summary by CodeRabbit

- **New Features**
	- Introduced a `ContentDialog` component for rendering and displaying Markdown content with clipboard functionality.
	- Added speech recognition capabilities with a new `useSpeechRecognition` function.
	- Enhanced the main chat interface using the new `dialog-content` component for better user interaction.
	- Added clickable chat tips for easier message sending.
	- Introduced a `TokenDialog` component for managing access tokens with validation.

- **Bug Fixes**
	- Improved error handling for API requests and enhanced feedback mechanisms.

- **Chores**
	- Updated dependencies to support new functionalities, including clipboard operations, Markdown processing, and syntax highlighting.

### CODE REVIEW

**Comment by CatsAndMice** [comment]
@chilingling 麻烦code  views下这个PR

**Review by CatsAndMice — APPROVED** [review]

File: packages/design-core/src/DesignPlugins.vue [hunk 1]
```diff
@@ -82,36 +82,56 @@
       </keep-alive>
     </div>
   </div>
-  <!--  AI 悬浮窗  -->
+
   <Teleport to="body">
+    <tiny-dialog-box v-model:visible="keyFormVisible" title="欢迎使用AI对话功能" >
```
Comment (chilingling) [inline packages/design-core/src/DesignPlugins.vue]: 这块属于 plugins/robot 的功能。麻烦顺带迁移到 plugins/robot 里面吧。

File: packages/plugins/robot/src/Main.vue [hunk 2]
```diff
@@ -171,26 +174,28 @@ export default {
...
-    const codeRules = `
-    请扮演一名前端开发专家，生成代码时遵从以下几条要求:
-###
-1. 只使用element-ui组件库完成代码编写
-2. 使用vue2技术栈
-3. 回复中只能有一个代码块
-4. el-table标签内不得出现el-table-column
-###
-  `
+//     const codeRules = `
```
Comment (chilingling) [inline packages/plugins/robot/src/Main.vue]: 不需要的注释可以删除掉

File: packages/plugins/robot/src/Main.vue [hunk 3]
```diff
@@ -347,7 +356,34 @@ export default {
...
+    const {
+      startRecognition,
+      stopRecognition,
+      recognizedText
+    } = useSpeechRecognition();
+
+    const speechStatus = ref(false)
+    const speechRecognition = () =>{
+      speechStatus.value = !speechStatus.value
+      console.log(speechStatus)
```
Comment (chilingling) [inline packages/plugins/robot/src/Main.vue]: `console.log` 不需要提交上来，调试完毕就可以删除

File: packages/plugins/robot/src/js/useDialogContent.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    clipboard.on('error', function (e) {
+      console.error('无法复制文本，请尝试手动复制。');
+    });
+  }
+
+  return {
+    md,
+    initClipboard,
+  };
+}
```
Comment (chilingling) [inline packages/plugins/robot/src/js/useDialogContent.js]: 文件末尾，请留一行空白行。下同

**Comment by chilingling** [comment]
另外，可以对 commit 加上 gpg 签名。
参考：https://docs.github.com/zh/authentication/managing-commit-signature-verification/signing-commits

File: packages/design-core/src/DesignPlugins.vue [hunk 5]
```diff
@@ -82,36 +82,56 @@
...
-  },
-  props: {
+  }, props: {
     renderPanel: {
       type: String
     }
-  },
-  emits: ['click', 'node-click'],
-  setup(props, { emit }) {
+  }, emits: ['click', 'node-click'], setup(props, { emit }) {
```
Comment (hexqi) [inline packages/design-core/src/DesignPlugins.vue]: 请使用prettier格式化代码

File: packages/plugins/robot/src/DialogContent.vue [hunk 6]
```diff
@@ -0,0 +1,63 @@
+<template>
+  <div v-html="renderedMarkdown" />
+</template>
+
+<script>
+import { ref, onMounted, watch, nextTick } from 'vue';
```
Comment (hexqi) [inline packages/plugins/robot/src/DialogContent.vue]: 需要格式化代码，保持跟项目风格统一(末尾不需要分号；)

File: packages/design-core/src/DesignPlugins.vue [hunk 7]
```diff
@@ -185,6 +205,48 @@ export default {
...
 
+    const keyFormVisible = ref(false)
+    const keyFormRef = ref(null)
+    const keyForm = ref({
+      accessToken: ''
+    })
+    const accessTokenReg = /^[A-Za-z0-9\-.]+$/
+    const accessTokenValidate = (rule, value, callback) => {
+      if (value.length > 100 || !accessTokenReg.test(value)) {
+        callback(new Error('参数错误，请输入小于100位的英文数字字符串'))
```
Comment (hexqi) [inline packages/design-core/src/DesignPlugins.vue]: UI调整与新增功能，麻烦在PR描述中补充上必要的描述文字/截图等信息

File: packages/plugins/robot/src/Main.vue [hunk 8]
```diff
@@ -171,26 +174,28 @@ export default {
...
+//     请扮演一名前端开发专家，生成代码时遵从以下几条要求:
+// ###
+// 1. 只使用element-ui组件库完成代码编写
+// 2. 使用vue2技术栈
+// 3. 回复中只能有一个代码块
+// 4. el-table标签内不得出现el-table-column
+// ###
+//   `
 
+    const codeRules = ''
```
Comment (hexqi) [inline packages/plugins/robot/src/Main.vue]: 去掉Prompt后，AI生成的代码能否在画布正常显示呢？

**Comment by Fleurxxx** [comment]
@CatsAndMice @chilingling @hexqi 我已经对代码进行了调整，麻烦重新code review一下

**Comment by hexqi** [comment]
> @CatsAndMice @chilingling @hexqi 我已经对代码进行了调整，麻烦重新code review一下

已经解决的意见可以点下Resolve Conversation

File: packages/design-core/src/DesignPlugins.vue (hunk 7 above)
Comment (hexqi) [reply packages/design-core/src/DesignPlugins.vue]: 设置access_token的对话框是点击哪里触发的呢, 似乎没有看到入口在哪里

Comment (Fleurxxx) [reply packages/design-core/src/DesignPlugins.vue]: 已修改，触发逻辑添加在robot/src/Main.vue这个文件中：
1. 在点击左侧ai插件图标时检查access_token并根据其存在与否显示对话框。
2. 在发送对话请求后，根据响应结果，若过期则清除access_token并触发对话框。

File: packages/plugins/robot/src/ContentDialog.vue [hunk 9]
```diff
@@ -0,0 +1,65 @@
+<template>
+  <div v-html="sanitize(renderedMarkdown)" />
```
Comment (CatsAndMice) [inline packages/plugins/robot/src/ContentDialog.vue]: sanitize 从哪引入，全局吗？

File: packages/plugins/robot/src/Main.vue [hunk 10]
```diff
@@ -55,34 +55,40 @@
...
-              <span>{{ item.content }}</span>
+              <dialog-content :markdownContent="item.content" />
             </div>
           </tiny-col>
         </tiny-row>
       </tiny-layout>
     </article>
     <article class="chat-tips">
-      <span>需要一个注册表单？</span>
+      <span @click="sendContent('需要一个注册表单？', true)">需要一个注册表单？</span>
```
Comment (CatsAndMice) [inline packages/plugins/robot/src/Main.vue:66]: 这里为啥有两行代码是一样的，而且触发的事件也一样。是不是多余了

Comment (hexqi) [reply packages/plugins/robot/src/Main.vue:66]: 可以删除一个

File: packages/plugins/robot/src/Main.vue [hunk 11]
```diff
@@ -172,20 +182,22 @@ export default {
...
-###
-1. 只使用element-ui组件库完成代码编写
-2. 使用vue2技术栈
-3. 回复中只能有一个代码块
-4. el-table标签内不得出现el-table-column
-###
-  `
-
+      从现在开始，请扮演一名前端专家。如果需要生成前端代码，代码中的所有组件必须使用 Vue 3 框架和 TinyVue 组件库进行编写。例如，如果你想使用按钮组件，应该使用 TinyVue 组件库中的 \`tiny-button\`。
+      以下是 TinyVue 组件库的文档，请通读并遵循其中的指导来生成代码：[TinyVue 组件库文档](https://opentiny.design/tiny-vue/zh-CN/os-theme/overview)
```
Comment (hexqi) [inline packages/plugins/robot/src/Main.vue]: 这一行Prompt有验证过吗，可能并不会有效果

File: packages/plugins/robot/src/Main.vue [hunk 12]
```diff
@@ -55,34 +55,39 @@
                   : 'chat-content-ai'
               ]"
             >
-              <span>{{ item.content }}</span>
+              <dialog-content :markdownContent="item.content" />
```
Comment (CatsAndMice) [inline packages/plugins/robot/src/Main.vue:58]: markdownContent 注意书写规范，应该用中划线

File: packages/plugins/robot/src/Main.vue [hunk 13]
```diff
@@ -172,20 +182,22 @@ export default {
...
+    // `
     // 在每一次发送请求之前，都把引入区块的内容，给放到第一条消息中
     // 为了不污染存储在localstorage里的用户的原始消息，这里进行了简单的对象拷贝
     // 引入区块不存放在localstorage的原因：因为区块是可以变化的，用户可能在同一个会话中，对区块进行了删除和创建。那么存放的数据就不是即时数据了。
     const getSendSeesionProcess = () => {
-      const sendProcess = { ...sessionProcess }
+      const accessToken = localStorage.getItem('accessToken')
+      const sendProcess = { ...sessionProcess, accessToken }
       const firstMessage = sendProcess.messages[0]
       firstMessage.content
```
Comment (CatsAndMice) [inline packages/plugins/robot/src/Main.vue:202]: 这一行是做什么的

File: packages/plugins/robot/src/TokenDialog.vue [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+}
+</script>
+
+<template>
+  <tiny-dialog-box :modal="false" v-model:visible="keyFormVisible" title="欢迎使用AI对话功能">
+    <tiny-form :model="keyForm" :rules="rules" ref="keyFormRef" label-width="0" class="demo-form">
+      <tiny-alert
+        :icon="TinyIconAssociation"
+        :closable="false"
+        description="当前AI大模型为使用文心一言：ERNIE-Bot-turbo"
```
Comment (hexqi) [inline packages/plugins/robot/src/TokenDialog.vue:87]: 插件中当前大模型是可以切换的，是动态的，建议不要在这里写死当前大模型的信息

File: packages/plugins/robot/src/TokenDialog.vue [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+    const closeKeyFormDialog = () => {
+      keyFormVisible.value = false
+      emit('dialog-status', false)
+    }
+
+    const submitKeyForm = () => {
+      keyFormRef.value.validate((valid) => {
+        if (valid) {
+          localStorage.setItem('accessToken', keyForm.value.accessToken)
```
Comment (hexqi) [inline packages/plugins/robot/src/TokenDialog.vue:61]: 保存accessToken也需要同时考虑模型切换，支持设置多个模型的token

**Review by CatsAndMice — APPROVED** [review]

File: packages/plugins/robot/src/Main.vue [hunk 16]
```diff
@@ -55,34 +55,39 @@
...
+            :class="{ 'microphone-svg': speechStatus }"
+            @click="speechRecognition"
+          ></svg-icon>
         </template>
       </tiny-input>
       <tiny-button @click="endContent">重新发起会话</tiny-button>
       <tiny-button @click="sendContent(inputContent, false)">发送</tiny-button>
     </footer>
   </div>
+  <token-dialog :dialog-visible="tokenDialogVisible" @dialog-status="getTokenDialogStatus"></token-dialog>
```
Comment (hexqi) [inline packages/plugins/robot/src/Main.vue:86]: 建议在插件中也添加一个设置图标，点击设置图标打开弹窗，修改access_token

File: packages/plugins/robot/src/TokenDialog.vue (hunk 14 above)
Comment (CatsAndMice) [reply packages/plugins/robot/src/TokenDialog.vue:87]: @hexqi 现在不是动态，计划是先固定一个AI模型，先跑通一个AI模型后再考虑接入更多的模型

File: packages/plugins/robot/src/TokenDialog.vue (hunk 15 above)
Comment (CatsAndMice) [reply packages/plugins/robot/src/TokenDialog.vue:61]: @hexqi 现在不是动态，计划是先固定一个AI模型，先跑通一个AI模型后再考虑接入更多的模型。在后端部分已经留好相应扩展代码

File: packages/plugins/robot/src/TokenDialog.vue (hunk 14 above)
Comment (hexqi) [reply packages/plugins/robot/src/TokenDialog.vue:87]: OK, 可以在后续PR中继续完善

File: packages/plugins/robot/src/Main.vue (hunk 16 above)
Comment (CatsAndMice) [reply packages/plugins/robot/src/Main.vue:86]: > 建议在插件中也添加一个设置图标，点击设置图标打开弹窗，修改access_token

@Fleurxxx 这个建议挺好的，方便用户修改

File: packages/plugins/robot/src/Main.vue (hunk 11 above)
Comment (Fleurxxx) [reply packages/plugins/robot/src/Main.vue]: 已验证，测试gif添加在上面对pr对描述中了

File: packages/plugins/robot/src/Main.vue [hunk 17]
```diff
@@ -232,9 +233,9 @@ export default {
           sessionProcess.displayMessages.push(respDisplayMessage)
           messages.value[messages.value.length - 1].content = originalResponse.choices?.[0]?.message.content
           setContextSession()
-          if (schema?.schema) {
-            createNewPage(schema.schema)
-          }
+          // if (schema?.schema) {
```
Comment (CatsAndMice) [inline packages/plugins/robot/src/Main.vue]: 注释的代码可以删除掉

Comment (Fleurxxx) [reply packages/plugins/robot/src/Main.vue]: 这是项目原来的代码，测试的时候注释掉了，因为还没有实现转为schema格式的代码会报错，现在取消注释重新推上去了

File: packages/plugins/robot/src/Main.vue [hunk 18]
```diff
@@ -8,6 +8,8 @@
           class="common-svg"
           @click="resizeChatWindow"
         ></svg-icon>
+        <SettingOutlined />
```
Comment (CatsAndMice) [inline packages/plugins/robot/src/Main.vue]: 注意书写规范，使用中划线

File: packages/plugins/robot/src/Main.vue [hunk 19]
```diff
@@ -172,20 +183,20 @@ export default {
...
-    请扮演一名前端开发专家，生成代码时遵从以下几条要求:
-###
-1. 只使用element-ui组件库完成代码编写
-2. 使用vue2技术栈
-3. 回复中只能有一个代码块
-4. el-table标签内不得出现el-table-column
-###
-  `
-
+     我想让你充当 Stackoverflow 的帖子。我将提出与编程有关的问题，你将回答答案是什么。我希望你只回答给定的答案，在没有足够的细节时写出解释。
```
Comment (hexqi) [inline packages/plugins/robot/src/Main.vue:186]: Prompt里面为啥要提stackoverflow呢，看起来有点奇怪

Comment (Fleurxxx) [reply packages/plugins/robot/src/Main.vue:186]: 我参考的是百度智能云里面prompt模板内容：https://console.bce.baidu.com/qianfan/prompt/template

Comment (Fleurxxx) [reply packages/plugins/robot/src/Main.vue:186]: 我需要去掉吗

Comment (hexqi) [reply packages/plugins/robot/src/Main.vue:186]: 先留着吧，后面可以多对比测试下，Prompt肯定还会不断调整的

**Review by hexqi — APPROVED** [review]

**Comment by hexqi** [comment]
还有部分commit没有签名
可以执行下面这条命令对从2c478ce开始的所有commit重新提交添加签名(当前只有最新提交的部分是有签名的)(执行完后:wq保存，git push -f提交)
`git rebase --exec 'GIT_COMMITTER_DATE="$(git log -1 --format=%at)" git commit --amend --no-edit -n -S' -i 2c478ce`

**Pull request merged**